certora/specs/rewardsPreserve.spec

```diff
@@ -0,0 +1,186 @@
+using DummyERC20_aTokenUnderlying as _DummyERC20_aTokenUnderlying 
+using RewardsControllerHarness as _RewardsController
+using TransferStrategyHarness as _TransferStrategyHarness
+
+
+methods
+{
+	// envfree
+	// -------
+	getUnclaimedRewards(address, address) returns (uint256) envfree
+    rewardTokens() returns (address[]) envfree
+	isRegisteredRewardToken(address) returns (bool) envfree
+    
+	// Getters from munged/harness
+    getRewardTokensLength() returns (uint256) envfree 
+    getRewardToken(uint256) returns (address) envfree
+
+	// AToken
+	_AToken.UNDERLYING_ASSET_ADDRESS() returns (address) envfree
+
+	// Reward token
+	_DummyERC20_rewardToken.balanceOf(address) returns (uint256) envfree
+	_DummyERC20_rewardToken.totalSupply() returns (uint256) envfree
+
+	// RewardsController
+    _RewardsController.getAvailableRewardsCount(address) returns (uint128) envfree
+    _RewardsController.getRewardsByAsset(address, uint128) returns (address) envfree
+	_RewardsController.getAssetListLength() returns (uint256) envfree
+	_RewardsController.getAssetByIndex(uint256) returns (address) envfree
+
+	// Permit
+	// ------
+    permit(address,address,uint256,uint256,uint8,bytes32,bytes32) => NONDET
+
+	// AToken
+	// ------
+    mint(address,address,uint256,uint256) returns (bool) => DISPATCHER(true)
+    burn(address,address,uint256,uint256) returns (bool) => DISPATCHER(true)
+}
+ 
+/**
+ * @title Single reward setup
+ * Setup the `StaticATokenLM`'s rewards so they contain a single reward token
+ * which is` _DummyERC20_rewardToken`.
+ */
+function single_RewardToken_setup() {
+	require isRegisteredRewardToken(_DummyERC20_rewardToken);
+	require getRewardTokensLength() == 1;
+	require getRewardToken(0) == _DummyERC20_rewardToken;
+}
+
+
+/**
+ * @title Single reward setup in RewardsController
+ * Sets (in `_RewardsController`) the first reward for `_AToken` as
+ * `_DummyERC20_rewardToken`.
+ */
+function rewardsController_reward_setup() {
+    require _RewardsController.getAvailableRewardsCount(_AToken) > 0;
+    require _RewardsController.getRewardsByAsset(_AToken, 0) == _DummyERC20_rewardToken;
+}
+
+
+/**
+ * @title Rewards claiming when sufficient rewards exist
+ * Ensures rewards are updated correctly after claiming, when there are enough
+ * reward funds.
+ *
+ * - Succeeds in: job-id=`1e47d3b29d4a4fae99ca7c001e7f65ae`
+ */
+rule rewardsConsistencyWhenSufficientRewardsExist() {
+	// Assuming single reward
+	single_RewardToken_setup();
+
+	// Create a rewards array
+	address[] _rewards;
+	require _rewards[0] == _DummyERC20_rewardToken;
+	require _rewards.length == 1;
+
+	env e;
+	require e.msg.sender != currentContract;  // Cannot claim to contract
+	uint256 rewardsBalancePre = _DummyERC20_rewardToken.balanceOf(e.msg.sender);
+	uint256 claimablePre = getClaimableRewards(e, e.msg.sender, _DummyERC20_rewardToken);
+
+	// Ensure contract has sufficient rewards
+	require _DummyERC20_rewardToken.balanceOf(currentContract) >= claimablePre;
```

Comment:
do we have a proof that
_DummyERC20_rewardToken.balanceOf(currentContract) >= sum(claimablePre)

Reply:
This is not always true, for example if the contract has unclaimed rewards in the `INCENTIVES_CONTROLLER`.
That's why in `_claimRewardsOnBehalf` we have the following code (`src/StaticATokenLM.sol:683--685`):
```solidity
      if (userReward > totalRewardTokenBalance) {
        totalRewardTokenBalance += collectAndUpdateRewards(address(rewards[i]));
      }
```
The code above collects the contracts' rewards if 
`_DummyERC20_rewardToken.balanceOf(currentContract) < claimablePre(user) `.

The strange thing is that even after collecting the rewards, the code does not assume it has enough for a single user, as shown in the code that follows (`src/StaticATokenLM.sol:687--690`):
```solidity
      if (userReward > totalRewardTokenBalance) {
        unclaimedReward = userReward - totalRewardTokenBalance;
        userReward = totalRewardTokenBalance;
      }
```

I tend to think that what is true is:
`_DummyERC20_rewardToken.balanceOf(currentContract) + collectAndUpdateRewards  >= sum(claimablePre)`.
Let's talk on how to prove that.